framework/include/base/MooseError.h

```diff
@@ -105,81 +100,79 @@
-        _deprecate_oss_                                                                                       \
-          << COLOR_YELLOW                                                                                     \
-          << "*** Warning, This code is deprecated, and likely to be removed in future library versions!\n"   \
-          << msg << '\n'                                                                                      \
-          << __FILE__ << ", line " << __LINE__ << ", compiled "                                               \
-          << LIBMESH_DATE << " at " << LIBMESH_TIME << " ***"                                                 \
-          << COLOR_DEFAULT                                                                                    \
-          << "\n\n";                                                                                          \
-        if (libMesh::global_n_processors() == 1)                                                              \
-          print_trace(_deprecate_oss_);                                                                       \
-        else                                                                                                  \
-          libMesh::write_traceout();                                                                          \
-        _console << _deprecate_oss_.str() << std::flush;                                                      \
-      );                                                                                                      \
-   } while (0)
+template <typename S, typename... Args>
+void
+mooseWarningStream(S & oss, Args... args)
+{
+  if (Moose::_warnings_are_errors)
+    mooseError(args...);
+
+  std::ostringstream ss;
+  mooseStreamAll(ss, args...);
+  std::string msg = mooseMsgFmt(ss.str(), "*** Warning ***", COLOR_YELLOW);
+  if (Moose::_throw_on_error)
+    throw std::runtime_error(msg);
+
+  oss << msg << std::flush;
+}
+
+template <typename S, typename... Args>
+void
+mooseInfoStream(S & oss, Args... args)
+{
+  mooseDoOnce(
+    {
+      std::ostringstream ss;
+      mooseStreamAll(ss, args...);
+      std::string msg = mooseMsgFmt(ss.str(), "*** Info ***", COLOR_CYAN);
+      oss << msg << std::flush;
+    }
+  );
+}
+
+template <typename S, typename... Args>
+void
+mooseDeprecatedStream(S & oss, Args... args)
+{
+  if (Moose::_deprecated_is_error)
+    mooseError("\n\nDeprecated code:\n", args...);
+
+  mooseDoOnce(
+    std::ostringstream ss;
+    mooseStreamAll(ss, args...);
+    std::string msg = mooseMsgFmt(ss.str(), "*** Warning, This code is deprecated and will be removed in future versions!", COLOR_YELLOW);
+    oss << msg;
+    ss.str("");
+    if (libMesh::global_n_processors() == 1)
+      print_trace(ss);
+    else
+      libMesh::write_traceout();
+    oss << ss.str() << std::endl;
+  );
+}
+
+template <typename... Args>
+void
+mooseWarning(Args... args)
```

Comment:
What const versions are you seeing?  If the passed in param is const, the function still works (const becomes part of the templated type(s)) and because we never perform operations on the value that wouldn't work with const.  I also couldn't get any of the variadic templating to work with const qualifiers in the signatures.  Maybe somebody with more templating foo than me could suggest how.